File: emergent-client/src/message.rs
//! Message types for the Emergent client library.

use mti::prelude::*;
use serde::{de::DeserializeOwned, Deserialize, Serialize};
use std::time::{SystemTime, UNIX_EPOCH};

/// Standard message envelope for all Emergent communications.
///
/// All messages in Emergent use this standard envelope format. Developers specify
/// a `message_type` string and put their domain data in the `payload` field.
///
/// # Example
///
/// ```rust
/// use emergent_client::EmergentMessage;
/// use serde_json::json;
///
/// let message = EmergentMessage::new("user.created")
///     .with_source("user_service")
///     .with_payload(json!({
///         "user_id": "u_12345",
///         "email": "user@example.com"
///     }));
/// ```
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct EmergentMessage {
    /// Unique message ID (MTI format: msg_<uuid_v7>).
    pub id: String,

    /// Message type for routing (e.g., "email.received", "timer.tick").
    pub message_type: String,

    /// Source client that published this message.
    pub source: String,

    /// Optional correlation ID for request-response or tracing.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub correlation_id: Option<String>,

    /// Optional causation ID (ID of message that triggered this one).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub causation_id: Option<String>,

    /// Timestamp when message was created (Unix ms).
    pub timestamp_ms: u64,

    /// User-defined payload (any serializable data).
    pub payload: serde_json::Value,

    /// Optional metadata for debugging, tracing, etc.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<serde_json::Value>,
}

impl EmergentMessage {
    /// Create a new message with the given type.
    ///
    /// Generates a unique ID and sets the current timestamp.
    #[must_use]
    pub fn new(message_type: &str) -> Self {
        Self {
            id: "msg".create_type_id::<V7>().to_string(),
            message_type: message_type.to_string(),
            source: String::new(),
            correlation_id: None,
            causation_id: None,
            timestamp_ms: current_timestamp_ms(),
            payload: serde_json::Value::Null,
            metadata: None,
        }
    }

    /// Set the source of this message.
    #[must_use]
    pub fn with_source(mut self, source: &str) -> Self {
        self.source = source.to_string();
        self
    }

    /// Set the payload of this message.
    #[must_use]
    pub fn with_payload(mut self, payload: impl Serialize) -> Self {
        self.payload = serde_json::to_value(payload).unwrap_or(serde_json::Value::Null);
        self
    }

    /// Set the correlation ID (for request-response patterns).
    #[must_use]
    pub fn with_correlation_id(mut self, id: &str) -> Self {
        self.correlation_id = Some(id.to_string());
        self
    }

    /// Set the causation ID (ID of the message that triggered this one).
    #[must_use]
    pub fn with_causation_id(mut self, id: &str) -> Self {
        self.causation_id = Some(id.to_string());
        self
    }

    /// Set optional metadata.
    #[must_use]
    pub fn with_metadata(mut self, metadata: impl Serialize) -> Self {
        self.metadata = Some(serde_json::to_value(metadata).unwrap_or(serde_json::Value::Null));
        self
    }

    /// Get the message ID.
    #[must_use]
    pub fn id(&self) -> &str {
        &self.id
    }

    /// Get the message type.
    #[must_use]
    pub fn message_type(&self) -> &str {
        &self.message_type
    }

    /// Get the source.
    #[must_use]
    pub fn source(&self) -> &str {
        &self.source
    }

    /// Get the raw payload value.
    #[must_use]
    pub fn payload(&self) -> &serde_json::Value {
        &self.payload
    }

    /// Deserialize the payload into a specific type.
    ///
    /// # Errors
    ///
    /// Returns an error if the payload cannot be deserialized into type `T`.
    pub fn payload_as<T: DeserializeOwned>(&self) -> Result<T, serde_json::Error> {
        serde_json::from_value(self.payload.clone())
    }

    /// Serialize the message to JSON bytes.
    ///
    /// # Errors
    ///
    /// Returns an error if serialization fails.
    pub fn to_json(&self) -> Result<Vec<u8>, serde_json::Error> {
        serde_json::to_vec(self)
    }

    /// Deserialize a message from JSON bytes.
    ///
    /// # Errors
    ///
    /// Returns an error if deserialization fails.
    pub fn from_json(data: &[u8]) -> Result<Self, serde_json::Error> {
        serde_json::from_slice(data)
    }

    /// Serialize the message to MessagePack bytes.
    ///
    /// # Errors
    ///
    /// Returns an error if serialization fails.
    pub fn to_msgpack(&self) -> Result<Vec<u8>, rmp_serde::encode::Error> {
        rmp_serde::to_vec(self)
    }

    /// Deserialize a message from MessagePack bytes.
    ///
    /// # Errors
    ///
    /// Returns an error if deserialization fails.
    pub fn from_msgpack(data: &[u8]) -> Result<Self, rmp_serde::decode::Error> {
        rmp_serde::from_slice(data)
    }
}

/// Get the current timestamp in milliseconds.
fn current_timestamp_ms() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| u64::try_from(d.as_millis()).unwrap_or(u64::MAX))
        .unwrap_or(0)
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn test_message_creation() {
        let msg = EmergentMessage::new("test.event")
            .with_source("test_source")
            .with_payload(json!({"key": "value"}));

        assert!(msg.id.starts_with("msg_"));
        assert_eq!(msg.message_type, "test.event");
        assert_eq!(msg.source, "test_source");
        assert!(msg.timestamp_ms > 0);
    }

    #[test]
    fn test_message_serialization() {
        let msg = EmergentMessage::new("test.event")
            .with_source("test")
            .with_payload(json!({"num": 42}));

        // Test JSON serialization
        let json_bytes = msg.to_json().unwrap();
        let from_json = EmergentMessage::from_json(&json_bytes).unwrap();
        assert_eq!(from_json.message_type, "test.event");

        // Test MessagePack serialization
        let msgpack_bytes = msg.to_msgpack().unwrap();
        let from_msgpack = EmergentMessage::from_msgpack(&msgpack_bytes).unwrap();
        assert_eq!(from_msgpack.message_type, "test.event");
    }

    #[test]
    fn test_payload_extraction() {
        #[derive(Debug, Deserialize, PartialEq)]
        struct TestPayload {
            count: u32,
            name: String,
        }

        let msg = EmergentMessage::new("test.event").with_payload(json!({
            "count": 42,
            "name": "test"
        }));

        let payload: TestPayload = msg.payload_as().unwrap();
        assert_eq!(payload.count, 42);
        assert_eq!(payload.name, "test");
    }

    #[test]
    fn test_message_tracing() {
        let original = EmergentMessage::new("request");
        let response = EmergentMessage::new("response")
            .with_causation_id(original.id())
            .with_correlation_id("corr_123");

        assert_eq!(response.causation_id.as_deref(), Some(original.id()));
        assert_eq!(response.correlation_id.as_deref(), Some("corr_123"));
    }
}

File: sdks/go/message.go
package emergent

import (
	"encoding/json"
)

// EmergentMessage is the universal message envelope for all Emergent communications.
type EmergentMessage struct {
	// ID is the unique message identifier (TypeID: msg_<uuid_v7>).
	ID MessageId `json:"id" msgpack:"id"`
	// MessageType is the message type for routing (e.g., "timer.tick").
	MessageType MessageType `json:"message_type" msgpack:"message_type"`
	// Source is the primitive that published this message.
	Source PrimitiveName `json:"source" msgpack:"source"`
	// CorrelationID is an optional correlation identifier for request-response patterns.
	CorrelationID string `json:"correlation_id,omitempty" msgpack:"correlation_id,omitempty"`
	// CausationID is an optional identifier of the message that triggered this one.
	CausationID string `json:"causation_id,omitempty" msgpack:"causation_id,omitempty"`
	// TimestampMs is the creation time in Unix milliseconds.
	TimestampMs Timestamp `json:"timestamp_ms" msgpack:"timestamp_ms"`
	// Payload is the user-defined message data.
	Payload any `json:"payload" msgpack:"payload"`
	// Metadata is optional debugging/tracing metadata.
	Metadata any `json:"metadata,omitempty" msgpack:"metadata,omitempty"`
}

// NewMessage creates a new EmergentMessage with the given type.
// Generates a unique ID and sets the timestamp to now.
func NewMessage(messageType string) (*EmergentMessage, error) {
	mt, err := NewMessageType(messageType)
	if err != nil {
		return nil, err
	}
	return &EmergentMessage{
		ID:          NewMessageId(),
		MessageType: mt,
		Source:      "unknown",
		TimestampMs: TimestampNow(),
	}, nil
}

// WithSource sets the source of this message and returns the message for chaining.
func (m *EmergentMessage) WithSource(source string) *EmergentMessage {
	m.Source = PrimitiveName(source)
	return m
}

// WithPayload sets the payload and returns the message for chaining.
func (m *EmergentMessage) WithPayload(payload any) *EmergentMessage {
	m.Payload = payload
	return m
}

// WithCorrelationID sets the correlation ID and returns the message for chaining.
func (m *EmergentMessage) WithCorrelationID(id string) *EmergentMessage {
	m.CorrelationID = id
	return m
}

// WithCausationID sets the causation ID and returns the message for chaining.
func (m *EmergentMessage) WithCausationID(id string) *EmergentMessage {
	m.CausationID = id
	return m
}

// WithCausationFromMessage sets the causation ID from a message ID and returns the message for chaining.
func (m *EmergentMessage) WithCausationFromMessage(msgID MessageId) *EmergentMessage {
	m.CausationID = string(msgID)
	return m
}

// WithMetadata sets the metadata and returns the message for chaining.
func (m *EmergentMessage) WithMetadata(metadata any) *EmergentMessage {
	m.Metadata = metadata
	return m
}

// HasStdoutPayload checks whether this message has an exec-source payload shape.
// Returns true if the payload is a map with a "stdout" key that is a string.
func (m *EmergentMessage) HasStdoutPayload() bool {
	payloadMap, ok := m.Payload.(map[string]any)
	if !ok {
		return false
	}
	_, ok = payloadMap["stdout"].(string)
	return ok
}

// UnwrapStdout extracts and parses the stdout field from an exec-source payload.
// If the payload is a map with a "stdout" string key, attempts to parse it as JSON.
// On success, the payload is replaced with the parsed value; on failure, the
// payload is replaced with the raw stdout string.
// If the payload doesn't match the exec-source shape, it is left unchanged.
// Returns self for chaining.
func (m *EmergentMessage) UnwrapStdout() *EmergentMessage {
	payloadMap, ok := m.Payload.(map[string]any)
	if !ok {
		return m
	}
	stdoutStr, ok := payloadMap["stdout"].(string)
	if !ok {
		return m
	}
	var parsed any
	if err := json.Unmarshal([]byte(stdoutStr), &parsed); err != nil {
		m.Payload = stdoutStr
	} else {
		m.Payload = parsed
	}
	return m
}

// PayloadAs unmarshals the payload into the given target.
// Uses JSON re-marshaling to support typed deserialization from map[string]any.
func (m *EmergentMessage) PayloadAs(target any) error {
	data, err := json.Marshal(m.Payload)
	if err != nil {
		return err
	}
	return json.Unmarshal(data, target)
}

// ToWire converts the message to wire format for serialization.
func (m *EmergentMessage) ToWire() map[string]any {
	wire := map[string]any{
		"id":           string(m.ID),
		"message_type": string(m.MessageType),
		"source":       string(m.Source),
		"timestamp_ms": uint64(m.TimestampMs),
		"payload":      m.Payload,
	}
	if m.CorrelationID != "" {
		wire["correlation_id"] = m.CorrelationID
	}
	if m.CausationID != "" {
		wire["causation_id"] = m.CausationID
	}
	if m.Metadata != nil {
		wire["metadata"] = m.Metadata
	}
	return wire
}

// MessageFromWire creates an EmergentMessage from a wire format map.
func MessageFromWire(wire map[string]any) *EmergentMessage {
	msg := &EmergentMessage{}

	if id, ok := wire["id"].(string); ok {
		msg.ID = MessageId(id)
	}
	if mt, ok := wire["message_type"].(string); ok {
		msg.MessageType = MessageType(mt)
	}
	if src, ok := wire["source"].(string); ok {
		msg.Source = PrimitiveName(src)
	}
	if cid, ok := wire["correlation_id"].(string); ok {
		msg.CorrelationID = cid
	}
	if cid, ok := wire["causation_id"].(string); ok {
		msg.CausationID = cid
	}
	if ts, ok := wire["timestamp_ms"].(uint64); ok {
		msg.TimestampMs = Timestamp(ts)
	} else if ts, ok := wire["timestamp_ms"].(int64); ok {
		msg.TimestampMs = Timestamp(ts)
	} else if ts, ok := wire["timestamp_ms"].(float64); ok {
		msg.TimestampMs = Timestamp(ts)
	}
	msg.Payload = wire["payload"]
	msg.Metadata = wire["metadata"]

	return msg
}
